How to Authenticate user (#13)
icure/medtech-docs: sdks/how-to/how-to-authenticate-a-user.md

| msgGtwSpecId         | Your iCure Message Gateway Identifier. See next section to know more about it                    |
| authProcessByEmailId | Identifier of the authentication by email process. See next section to know more about it        |
| authProcessBySmsId   | Identifier of the authentication by SMS process. See next section to know more about it          |

As Daenaerys needs to be identified as a patient, you will have to provide the `patientAuthProcessByEmailId` as a 
authProcessByEmailId and `patientAuthProcessBySmsId` as a authProcessBySmsId. 

:::info

If Daenaerys was a doctor, you would instead provide the `hcpAuthProcessByEmailId` as a
authProcessByEmailId and `hcpAuthProcessByEmailId` as a authProcessBySmsId.

:::

## Starting the authentication process
In iCure, you can authenticate either by email, either by SMS.
Therefore, you will have to ask Daenaerys to provide at least an email or a mobile phone when authenticating.

You will also have to implement the ReCAPTCHA mechanism and provide us the corresponding to found key during the startAuthentication

> process.

Author: Bad formulation : I wanted to refer the recaptcha computed score

:::info

Check the official [reCAPTCHA v3 documentation](https://developers.google.com/recaptcha/docs/v3) for more information.
Also, do not forget to contact the iCure team to get our ReCAPTCHA SiteKey that you will need to implement the reCAPTCHA

:::

<!-- file://code-samples/how-to/authenticate-user/index.mts snippet:Start Authentication Process By Email-->
```typescript

> const authProcess = await anonymousApi.authenticationApi.startAuthentication(

Author: I prefer letting the `startAuthentication` / `completeAuthentication` methods as is. 
From developer point of view, the registration / login process is the same : Ensuring the user is who he meant to be. 
That way, developers can call the service without splitting the behaviour between a registration or a login if they don't want : We take care of it (Just like Authentication through Google / Facebook, ... on mobile apps). 
However, I greatly simplified the use of startAuthentication / completeAuthentication so that developers don't have to provide non useful information for the login process anymore

  authProcess!,
  validationCode,
  [userPrivateKey, userPublicKey],
  () => undefined
)

const authenticatedApi = authenticationResult.medTechApi

console.log(`Your new user id: ${authenticationResult.userId}`)
console.log(`Database id where new user was created: ${authenticationResult.groupId}`)
console.log(`Your initialised MedTechAPI: ***\${authenticatedApi}***`)
console.log(`RSA keypair of your new user: ***\${authenticationResult.keyPair}***`)
console.log(`Token created to authenticate your new user: ***\${authenticationResult.token}***`)
```

As a result, you receive : 
- The MedTechApi instance to use for Daenaerys (properly initialised);
- The `userId`, identifying Daenaerys user uniquely;
- The `groupId`, identifying the database in which Daenaerys was created;
- The `keyPair`, the RSA keypair you created for Daenaerys; 
- The `token`, the time-limited token created for Daenaerys, to authenticate her; 

Make sure to save these elements to be able to authenticate Daenaerys again when she'll come back on your app.

<!-- file://code-samples/how-to/authenticate-user/index.mts snippet:Save credentials-->
```typescript

> // saveSecurely does not exist : Use your own way of storing the following data securely

Reviewer: ```suggestion
// saveSecurely does not exist: Use your own way of storing the following data securely
```

// One option is to put these elements into the localStorage
saveSecurely(userEmail, authenticationResult.token, authenticationResult.userId, authenticationResult.groupId, authenticationResult.keyPair)
```

> 

Reviewer: I think it would be a good idea to add here that if the private key is stored only in the browser local storage and if the user deletes is local browser data they may lose access to data.
It would be a good idea to suggest our user to add request to their app to tell their user to store the private key somewhere in their pc.
```suggestion

:::danger

If you are building a web app and store the private key only in your user's browser local storage you should
consider that if the user deletes their browser data they will lose access to the data they created in iCure. 
After completing user registration it might be a good idea to ask your user to store their private key
in a safe place in their filesystem, possibly encrypting it with a password.
Make sure your users understand they should never share this file with anyone.

:::
```

  "labels": {},
  "codes": {},
  "names": [
    {
      "firstNames": [
        "John"
      ],
      "prefix": [],
      "suffix": [],
      "lastName": "Snow",
      "text": "Snow John",
      "use": "official"
    }
  ],
  "languages": [],
  "addresses": [],
  "mergedIds": {},
  "active": true,
  "deactivationReason": "none",
  "partnerships": [],
  "patientHealthCareParties": [],
  "patientProfessions": [],
  "parameters": {},
  "properties": {},
  "rev": "1-8e3ad0d7e3179188dcd95f186f78b68d",
  "created": 1664552695128,
  "modified": 1664552695128,
  "author": "3363719b-579e-4640-ac62-13e608e69395",
  "responsible": "d6c8dbc7-eaa8-4c95-b9b3-920fb70ce59b",
  "firstName": "John",
  "lastName": "Snow",
  "gender": "male",
  "birthSex": "unknown",
  "personalStatus": "unknown",
  "note": "Winter is coming",
  "systemMetaData": {
    "hcPartyKeys": {},
    "privateKeyShamirPartitions": {},
    "secretForeignKeys": [],
    "cryptedForeignKeys": {},
    "delegations": {
      "d6c8dbc7-eaa8-4c95-b9b3-920fb70ce59b": {}
    },
    "encryptionKeys": {
      "d6c8dbc7-eaa8-4c95-b9b3-920fb70ce59b": {}
    },
    "aesExchangeKeys": {},
    "transferKeys": {}
  }
}
```

</details>

> But what do you have to do when Daenaerys will need to login again in a few days ?

Reviewer: ```suggestion
But what do you have to do when the authentication token of Daenaerys expires and she needs to login again?
```

# Login a user

> In iCure, the login flow is extremely similar to the registration.
> Once Daenaerys token is expired, she will need to identify herself again by starting a new authentication process. 

Reviewer: ```suggestion
In iCure, the login flow is similar to the registration.
Once Daenaerys's token is expired, she will need to authenticate again to iCure by starting the login process. 
```

As Daenaerys is not authenticated anymore, you have to create a new AnonymousMedTechApi instance. 

<!-- file://code-samples/how-to/authenticate-user/index.mts snippet:Login-->
```typescript
const anonymousApiForLogin = await new AnonymousMedTechApiBuilder()
  .withICureUrlPath(iCureUrl)
  .withCrypto(webcrypto as any)
  .withMsgGtwUrl(msgGtwUrl)
  .withMsgGtwSpecId(specId)
  .withAuthProcessByEmailId(authProcessByEmailId)
  .withAuthProcessBySmsId(authProcessBySmsId)
  .build()

const authProcessLogin = await anonymousApiForLogin.authenticationApi.startAuthentication(
  masterHcpId,
  'Daenerys',
  'Targaryen',
  recaptcha,
  false,
  userEmail // The email address used for user registration
)
```

<details>
    <summary>Output</summary>

```json

```

</details>

Daenaerys then receives a new validation code by email.

> As you already created an RSA keypair for her, you just need to provide it into the `completeAuthentication` service, 
> getting it back from where you stored it previously. 

Reviewer: ```suggestion
Since you already created an RSA keypair for her, you just need to retrieve it from where you stored it previously
and provide it to the `completeAuthentication` method. 
```

<!-- file://code-samples/how-to/authenticate-user/index.mts snippet:Complete login authentication process-->
```typescript
const loginResult = await anonymousApiForLogin.authenticationApi.completeAuthentication(
  authProcessLogin!,
  validationCodeForLogin,
  [userPrivateKey, userPublicKey],
  () => undefined
)

console.log(`Your new user id: ${authenticationResult.userId}`)
console.log(`Database id where new user was created: ${authenticationResult.groupId}`)
console.log(`Your new initialised MedTechAPI: ***\${loginResult.medTechApi}***`)
console.log(`RSA keypair of your user stays the same: ***\${authenticationResult.keyPair}***`)

> console.log(`The token of your user will change: ***\${authenticationResult.token}***`)

Reviewer: shouldn't you log `loginResult.stuff` instead of `authenticationResult.stuff`?

  "labels": {},
  "codes": {},
  "names": [
    {
      "firstNames": [
        "John"
      ],
      "prefix": [],
      "suffix": [],
      "lastName": "Snow",
      "text": "Snow John",
      "use": "official"
    }
  ],
  "languages": [],
  "addresses": [],
  "mergedIds": {},
  "active": true,
  "deactivationReason": "none",
  "partnerships": [],
  "patientHealthCareParties": [],
  "patientProfessions": [],
  "parameters": {},
  "properties": {},
  "rev": "1-8e3ad0d7e3179188dcd95f186f78b68d",
  "created": 1664552695128,
  "modified": 1664552695128,
  "author": "3363719b-579e-4640-ac62-13e608e69395",
  "responsible": "d6c8dbc7-eaa8-4c95-b9b3-920fb70ce59b",
  "firstName": "John",
  "lastName": "Snow",
  "gender": "male",
  "birthSex": "unknown",
  "personalStatus": "unknown",
  "note": "Winter is coming",
  "systemMetaData": {
    "hcPartyKeys": {},
    "privateKeyShamirPartitions": {},
    "secretForeignKeys": [],
    "cryptedForeignKeys": {},
    "delegations": {
      "d6c8dbc7-eaa8-4c95-b9b3-920fb70ce59b": {}
    },
    "encryptionKeys": {
      "d6c8dbc7-eaa8-4c95-b9b3-920fb70ce59b": {}
    },
    "aesExchangeKeys": {},
    "transferKeys": {}
  }
}
```

</details>

> Last thing you need to know is what to do when Daenaerys credentials are still valid. 
> You don't want her to go through an email / SMS process each time she uses your app. Therefore, you have to re-use her 
> existing token and RSA keypair. 

Reviewer: ```suggestion
The last thing you need to know is what to do when Daenaerys credentials are still valid: if you saved 
the result from the login (or registration) process, the token may still be valid the next time Daenaerys
accesses your application.
In this case you can reuse the existing token and avoid having Daenaerys go through the login process
with email or SMS OTP every time she opens your application.
```

# Reusing existing credentials

> Each time you complete an authentication process (Register or login), you have to save the credentials you receive
> in a secured place.
> We symbolized it through the `saveSecurely` method.

Reviewer: ```suggestion
Each time you complete the registration or login process, you can save the credentials you receive
in a secured place.
We symbolised it through the `saveSecurely` method.
```

<!-- file://code-samples/how-to/authenticate-user/index.mts snippet:Save credentials-->
```typescript
// saveSecurely does not exist : Use your own way of storing the following data securely
// One option is to put these elements into the localStorage
saveSecurely(userEmail, authenticationResult.token, authenticationResult.userId, authenticationResult.groupId, authenticationResult.keyPair)
```

> First thing you have to do is to retrieve Daenaerys credentials and her RSA Keypair 

Reviewer: ```suggestion
The first thing you have to do is to retrieve Daenaerys credentials and her RSA Keypair 
```

<!-- file://code-samples/how-to/authenticate-user/index.mts snippet:Get back credentials-->
```typescript
// getBackCredentials does not exist : Use your own way of storing the following data securely
// One option is to get them back from the localStorage
const { login, token, pubKey, privKey } = getBackCredentials()
```

> And then, initialize a MedTechApi, authenticating Daenaerys directly. 

Reviewer: ```suggestion
And then, initialise a MedTechApi, authenticating Daenaerys directly. 
```

  "codes": {},
  "names": [
    {
      "firstNames": [
        "John"
      ],
      "prefix": [],
      "suffix": [],
      "lastName": "Snow",
      "text": "Snow John",
      "use": "official"
    }
  ],
  "languages": [],
  "addresses": [],
  "mergedIds": {},
  "active": true,
  "deactivationReason": "none",
  "partnerships": [],
  "patientHealthCareParties": [],
  "patientProfessions": [],
  "parameters": {},
  "properties": {},
  "rev": "1-8e3ad0d7e3179188dcd95f186f78b68d",
  "created": 1664552695128,
  "modified": 1664552695128,
  "author": "3363719b-579e-4640-ac62-13e608e69395",
  "responsible": "d6c8dbc7-eaa8-4c95-b9b3-920fb70ce59b",
  "firstName": "John",
  "lastName": "Snow",
  "gender": "male",
  "birthSex": "unknown",
  "personalStatus": "unknown",
  "note": "Winter is coming",
  "systemMetaData": {
    "hcPartyKeys": {},
    "privateKeyShamirPartitions": {},
    "secretForeignKeys": [],
    "cryptedForeignKeys": {},
    "delegations": {
      "d6c8dbc7-eaa8-4c95-b9b3-920fb70ce59b": {}
    },
    "encryptionKeys": {
      "d6c8dbc7-eaa8-4c95-b9b3-920fb70ce59b": {}
    },
    "aesExchangeKeys": {},
    "transferKeys": {}
  }
}
```

</details>

# What's next ? 

> Some specific use cases can bring you some questions : What happens if Daenaerys lost her RSA Keypair ? 
> What happens if Daenaerys would like to start your app on another terminal ? 
> 
> Those questions will be answered in the [In-Depth Explanations](../explanations.md), in a future version of the documentation. 

Reviewer: ```suggestion
Some specific use cases can bring you some questions: what happens if Daenaerys lost her RSA Keypair? 
What happens if Daenaerys would like to start your app on another device? 

Those questions will be answered in the [In-Depth Explanations](../explanations.md)
```
